### ISSUE TITLE
Sylius\Component\Core\Model\Product conflicts with custom variant class in configuration

### ISSUE THREAD

**Issue by msumme** [Sylius/Sylius#2783]
When trying to run `sylius:install:sample-data` with a custom product class, I ran into the following:

[Doctrine\ORM\ORMInvalidArgumentException]
  Expected value of type "Doctrine\Common\Collections\Collection|array" for association field "AppBundle\Entity\Seatcover#$variants", got "Sylius\Component\Core\Model\ProductVariant" instead.

I tracked this down, and it was caused because my custom ProductVariant class wasn't being used.

I dug around, and figured out that Sylius\Component\Core\Model\Product sets the variant in __construct(), thus hard-coding the class and causing the doctrine exception.

```
 /**
     * Constructor.
     */
    public function __construct()
    {
        parent::__construct();

        $this->setMasterVariant(new ProductVariant());
        $this->taxons = new ArrayCollection();

        $this->variantSelectionMethod = self::VARIANT_SELECTION_CHOICE;
    }
```

**Comment by msumme** [comment]
Workaround: change your custom product class:
```
 /**
     * Constructor.
     */
    public function __construct()
    {
        parent::__construct();
        // override master variant to be the correct class
        $this->removeVariant($this->getMasterVariant());
        $this->setMasterVariant(new ProductVariant());
    }
```

**Comment by michalmarcinkowski** [comment]
Unfortunately that's the workaround you have to do in order to use different class. It's going to change in the coming weeks :)

**Comment by coudenysj** [comment]
@ michalmarcinkowski so what is the proper solution for this?

**Comment by pjedrzejewski** [comment]
This can be fixed with same approach #3480 thanks to factories.

**Comment by michalmarcinkowski** [comment]
Fixed in #3744
